src/Lean/Elab/Tactic/DiscrTreeKey.lean

```diff
@@ -0,0 +1,61 @@
+/-
+Copyright (c) 2024 Matthew Robert Ballard. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Tomas Skrivan, Matthew Robert Ballard
+-/
+prelude
+import Init.Tactics
+import Lean.Elab.Command
+import Lean.Meta.Tactic.Simp.SimpTheorems
+
+namespace Lean.Elab.Tactic.DiscrTreeKey
+
+open Lean.Meta DiscrTree
+open Lean.Elab.Tactic
+open Lean.Elab.Command
+
+private def mkKey (e : Expr) : MetaM (Array Key) := do
+  let (_, _, type) ← withReducible <| forallMetaTelescopeReducing e
+  let type ← whnfR type
```

Comment:
It is not good idea to return `type` from a telescope as it will end up with fvars that are not in the context anymore. However, I'm not sure if it matters for `DiscrTree.mkPath` or not.

I would write it as
```
private def mkKey (e : Expr) : MetaM (Array Key) :=
  withReducible <| forallTelescopeReducing e fun _ type => do
    let type ← whnfR type
    match type.eq? with
    | some (_, lhs, _) => mkPath lhs simpDtConfig
    | none => throwError "unexpected kind of 'simp' theorem{indentExpr type}"
```

Reply:
Ah, yes, of course, keep everything in the `forallTelescopeReducing`. My bad.

Reply:
I did notice some dropped fvars in the examples so I will try this out. Thanks!

Reply:
We end up with loose `bvar`'s without using `forAllMetaTelescopingReducing` to convert them into metavariables. Using this seems existing pattern, see `Lean.Meta.Simp.mkSimpTheoremCore` (private). So the output with this is consistent with the output of `set_option diagnostics`.

Reply:
Odd, but fine.